rust_backend/src/game.rs

```diff
@@ -0,0 +1,224 @@
+
+        if self.players.len() == 2 {
+            self.state = GameState::InProgress;
+        }
+
+        Ok(())
+    }
+
+    pub fn handle_msg(&mut self, msg: &ClientActorMessage) -> Result<(), AppError> {
+        if self.state == GameState::GameOver || self.state == GameState::NotStarted {
+            return Err(AppError {
+                message: format!("Game state is {:?}", self.state),
+                error_type: AppErrorType::GameStateError,
+            });
+        }
+
+        let opponent_id = self.get_opponent_id(&msg.id)?;
+        match msg.msg {
+            Msg::Hit => self.handle_attack(&msg.id, &opponent_id)?,
+            Msg::Wait => {},
+            _ => self.handle_movement(&msg.id, &msg.msg)?,
+        }
+
+        Ok(())
+    }
+
+    fn get_opponent_id(&self, player_id: &Uuid) -> Result<Uuid, AppError> {
+        for (uuid, _) in &self.players {
+            if uuid != player_id {
+                return Ok(uuid.clone());
+            }
+        }
+
+        Err(AppError {
+            message: "Opponent player not found".to_string(),
+            error_type: AppErrorType::PlayerNotFoundError,
+        })
+    }
+
+    pub fn handle_movement(&mut self, player_id: &Uuid, direction: &Msg) -> Result<(), AppError> {
+        let player = self.get_player_mut(player_id)?;
+
+        if direction == &Msg::Left {
+            player.go_left();
+        } else {
+            player.go_right();
+        }
+
+        Ok(())
+    }
+
+    pub fn handle_attack(&mut self, attacker_id: &Uuid, def_id: &Uuid) -> Result<(), AppError> {
+        if self.in_attack_zone(attacker_id, def_id)? {
+            let attacker = self.get_player(attacker_id)?;
+            let damage = attacker.damage;
+            let def_player = self.get_player_mut(def_id)?;
+            def_player.take_damage(damage);
+            if !def_player.is_alive() {
+                self.state = GameState::GameOver;
+                self.winner = Winner::Id(attacker_id.clone());
+            }
+        }
+
+        Ok(())
+    }
+
+    fn in_attack_zone(&self, attacker_id: &Uuid, def_id: &Uuid) -> Result<bool, AppError> {
+        let attacker = self.get_player(attacker_id)?;
+        let def_player = self.get_player(def_id)?;
+
+        let body_radius = def_player.body.width as f64 / 2f64;
+        let left_border = def_player.position.x - body_radius;
+        let right_border = def_player.position.x - body_radius;
+
+        let left_attack_zone = attacker.position.x - attacker.attack_zone as f64;
+        let right_attack_zone = attacker.position.x + attacker.attack_zone as f64;
+
+        return if self.in_range(&right_attack_zone, &left_border, &right_border) {
+            Ok(true)
+        } else if self.in_range(&left_attack_zone, &left_border, &right_border) {
+            Ok(true)
+        } else {
+            Ok(false)
+        };
+    }
+
+    fn in_range(&self, val: &f64, left_border: &f64, right_border: &f64) -> bool {
```

Comment:
there is no need to take primitive copy types by reference, just take them by value